### PR TITLE
Move ApiFeature-specific NodeSettings into ApiSettings class

### PR DESCRIPTION
**Pull request by quantumagi**
This PR adds a settings class to the Api Feature to be consistent with the other features.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        /// <summary>
+        /// Loads the storage related settings from the application configuration.
+        /// </summary>
+        /// <param name="nodeSettings">Application configuration.</param>
+        public virtual void Load(NodeSettings nodeSettings)
+        {
+            var config = nodeSettings.ConfigReader;
+
+            this.ApiUri = config.GetOrDefault("apiuri", new Uri(
+                $"http://localhost:{ (nodeSettings.Network.ToString().StartsWith("Stratis") ? 37221 : 37220) }"));
```
Comment (mikedennis) [inline src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: There is a network extension method Network.IsBitcoin() you could use that does the string comparison.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /// <summary>
+    /// Configuration related to storage of transactions.
+    /// </summary>
+    public class ApiSettings
+    {
+        /// <summary>URI to node's API interface.</summary>
+        public Uri ApiUri { get; set; }       
+
+        private Action<ApiSettings> callback = null;
```
Comment (mikedennis) [inline src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: I'm not sure if you need this or not. I haven't really been able to figure out what exactly the usage for this is. I know when I moved ConsensusSettings into it's own class I didn't bring this callback stuff along and it has had no ill effects. I recommend if you have no code consuming then get rid of it.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            : this(callback)
+        {
+            this.Load(nodeSettings);
+        }
+
+        /// <summary>
+        /// Loads the storage related settings from the application configuration.
+        /// </summary>
+        /// <param name="nodeSettings">Application configuration.</param>
+        public virtual void Load(NodeSettings nodeSettings)
```
Comment (mikedennis) [inline src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: Not sure why this is virtual? Also can't this be private as it's only used in constructor?

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: The callback provides a way to provide startup options - e.g. `.UseApi(s => s.ApiUri = "http://someuri")`

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 3 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: It is used in the ApiFeature to load the settings from the NodeSettings.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: Changed.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+        /// <summary>
+        /// Loads the API related settings from the application configuration.
+        /// </summary>
+        /// <param name="nodeSettings">Application configuration.</param>
+        public void Load(NodeSettings nodeSettings)
+        {
+            TextFileConfiguration config = nodeSettings.ConfigReader;
+
+            this.ApiUri = config.GetOrDefault("apiuri", new Uri($"http://localhost:{ (nodeSettings.Network.IsBitcoin() ? 37220 : 37221) }"));
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: personal opinion - if any more complex constructs are used in interpolated strings, the whole purpose of the interpolated string is thrown out of the window. if you agree, you might want to introduce a variable `port` here and then write

`$"http://localhost:{port}"`

which is very readable (and i think that was the goal with interpolated strings)

alternatively, you can do very well without port variable with old string format:
`string.Format("http://localhost:{0}", nodeSettings.Network.IsBitcoin() ? 37220 : 37221)`

which is imho also more readable than the interpolated string you have there now

but i would prefer the variable with interpolation over string format

File: src/Stratis.Bitcoin.Features.Api/Program.cs [hunk 5]
```diff
@@ -16,11 +16,13 @@ public static IWebHost Initialize(IEnumerable<ServiceDescriptor> services, FullN
         {
             Guard.NotNull(fullNode, nameof(fullNode));
 
+            var ApiUri = fullNode.NodeService<ApiSettings>().ApiUri;
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Features.Api/Program.cs]: unvar and small `a`piUri

File: src/Stratis.Bitcoin.IntegrationTests/APITests.cs [hunk 6]
```diff
@@ -65,7 +65,7 @@ public void CanGetGeneralInfoViaAPI()
             try
             {
                 var fullNode = this.apiTestsFixture.stratisPowNode.FullNode;
-                var ApiURI = fullNode.Settings.ApiUri;
+                var ApiURI = fullNode.NodeService<ApiSettings>().ApiUri;
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.IntegrationTests/APITests.cs]: `apiUri` i believe is the correct name per our coding style (also many times below)

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 3 above)
Comment (mikedennis) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: Ok I see that but is that necessary as it seems like Load is also called in the constructor when this class in injected. So is the net effect Load is called twice?

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 2 above)
Comment (mikedennis) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: Ah yes that's kind of cool. I wonder if I should also add that feature to consensus settings as it currently doesn't have that support.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 3 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: `public ApiSettings(NodeSettings nodeSettings, Action<ApiSettings> callback = null)' is not currently used so I have removed it. `Load` is called from the ApiFeature constructor.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 4 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: Changed to use `$"http://localhost:{port}"` to improve readability.

File: src/Stratis.Bitcoin.Features.Api/Program.cs (hunk 5 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Api/Program.cs]: Changed to `Uri apiUri = fullNode.NodeService<ApiSettings>().ApiUri;`

File: src/Stratis.Bitcoin.IntegrationTests/APITests.cs (hunk 6 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.IntegrationTests/APITests.cs]: Fixed the case introduced by legacy commit 292a2c8c:

`var apiURI = fullNode.NodeService<ApiSettings>().ApiUri;`

**Comment by bokobza** [comment]
I'm having a look at this as well.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    /// <summary>
+    /// Configuration related to the API interface.
+    /// </summary>
+    public class ApiSettings
+    {
+        /// <summary>URI to node's API interface.</summary>
+        public Uri ApiUri { get; set; }
+
+        /// <summary>The callback used to override/constrain/extend the settings provided by the Load method.</summary>
+        private Action<ApiSettings> callback = null;
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: I don't think this `= null` is necessary.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+        /// <summary>
+        /// Constructs this object.
+        /// </summary>
+        public ApiSettings()
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: is this needed?

File: src/Stratis.Bitcoin.Features.Api/Program.cs (hunk 5 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Api/Program.cs]: I think this hasn't been changed?

File: src/Stratis.Bitcoin.IntegrationTests/APITests.cs [hunk 9]
```diff
@@ -65,14 +65,14 @@ public void CanGetGeneralInfoViaAPI()
             try
             {
                 var fullNode = this.apiTestsFixture.stratisPowNode.FullNode;
-                var ApiURI = fullNode.Settings.ApiUri;
+                var apiURI = fullNode.NodeService<ApiSettings>().ApiUri;
```
Comment (bokobza) [inline src/Stratis.Bitcoin.IntegrationTests/APITests.cs]: can you please change this variable name to `apiUri`?

File: src/Stratis.Bitcoin.Features.Api/Program.cs (hunk 5 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Api/Program.cs]: Program.cs was omitted from the commit. Added.

File: src/Stratis.Bitcoin.IntegrationTests/APITests.cs (hunk 9 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.IntegrationTests/APITests.cs]: Program.cs was omitted from the commit. Added.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 8 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: Removed.

File: src/Stratis.Bitcoin.Features.Api/ApiSettings.cs (hunk 7 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Api/ApiSettings.cs]: Removed.

**Pull request merged**